### ISSUE TITLE
Replacement for express.json's "verify" functionality

### ISSUE THREAD

**Issue by zbjornson** [fastify/fastify#1562]
## 🚀 Feature Proposal

I'm moving an app from Express, where I have this code that uses a shared secret to verify the integrity of the request body (coming from GitHub web hooks):

```js
const parser = express.json({
	limit: "2mb",
	verify(req, res, buf, encoding) {
		const digest = crypto.createHmac("sha1", secret).update(buf).digest("hex");
		const signature = `sha1=${digest}`;
		const expect = req.headers["x-hub-signature"];
		if (expect !== signature)
			throw new Error("Signature verification failed");
	}
});

app.post("somepath", parser, (req, res, next) => {...});
// Note that there is *no* app.use(express.json) before this
```

The only way I've figured out how to do it with Fastify is to replace the default json parser:

```js
fastify.addContentTypeParser("application/json", {parseAs: "buffer"}, (req, body, done) => {
	const digest = crypto.createHmac("sha1", webhookSecret).update(body).digest("hex");
	req.signature = `sha1=${digest}`; // either attach it here
	// c/p from the default parser:
	try {
		var json = bourne.parse(body.toString("utf8"), {protoAction: "error"});
	} catch (err) {
		err.statusCode = 400;
		return done(err, undefined);
	}
	json.signature = `sha1=${digest}`; // or here ... or put it in a WeakMap<req, signature>
	done(null, json);
});
```

and then in my `fastify.route` handler, compare the header with the digest.

It'd be really nice if I could do this:
```js
fastify.route({
  verify(req, reply, buffer) { /* verify the buffer here, throw if bad */ }
  // ...
})
```

Might be a wontfix -- it's obviously possible to do this currently, I just don't love having to copy/paste the actual parser and having it apply to all routes. (I could put an `if (req.url.matches(...))` in the parser.)

## Motivation & Example

(above)

**Comment by Eomm** [comment]
Did you explore the [fastify hooks](https://github.com/fastify/fastify/blob/master/docs/Hooks.md)?

I think that `preParsing` or `preHandler` could fit for your needs

**Comment by zbjornson** [comment]
Yeah, it doesn't seem quite perfect though.

```js
fastify.addHook("preParsing", (req, reply, next) => {
	const buffs = [];
	req.raw.on("data", d => buffs.push(d));
	req.raw.on("end", () => {
		const buffbod = Buffer.concat(buffs);
		// calculate integrity here
	});
	next();
});
```

* Both this hook and the contentTypeParser are concatenating the stream (extra work and memory)
* There's no declarative way to apply it to a single route; need to use an `if (req.url.matches...)` guard in the hook.
* Need to manually check the request body size is less than the size limit.

Replacing the default "application/json" parser alleviates the 1st and 3rd issue, but again comes at the cost of having to duplicate the parser code itself.

**Comment by mcollina** [comment]
> it's obviously possible to do this currently, I just don't love having to copy/paste the actual parser and having it apply to all routes. 

I don't understand this. Plugins are made so you can easily reuse functionality across your codebase.

> Both this hook and the contentTypeParser are concatenating the stream (extra work and memory)

The source stream is always consumed only once. As you are parsing JSON, you need the full content anyway. Minus overhead and "more code", there is no extra work and memory being consumed.

---

Unfortunately I do not see how a `verify`  option can fit our model of content type parsers and hooks. However, there are a bunch of things we can do to make it easier:

* [ ] Add `preParsing` option to `.route()` , similar to `preValidation` (might just be that's not documented).
* [ ] Add an option to have per-route content type parsers.
* [ ] Expose the default JSON parser so that's reusable.

Wdyt?

**Comment by zbjornson** [comment]
(Sorry for the delay, was out sick.)

Sorry I was unclear on all that.

> I just don't love having to copy/paste the actual parser

I meant that the `try/catch` block in my OP is copied from here: https://github.com/fastify/fastify/blob/5de74406dbff5cba90516f6c9f499ea9d5e70e1b/lib/contentTypeParser.js#L200-L206

> and having it apply to all routes.

and here I meant that all `application/json` requests will use that parser; I can't have it only run on specific routes.

> The source stream is always consumed only once. As you are parsing JSON, you need the full content anyway. Minus overhead and "more code", there is no extra work and memory being consumed.

I retract this; my hook should be creating the HMAC once and calling `update()` on each `"data"` event to avoid buffering the stream into memory.

---

Adding `preParsing` to `route` would I think solve this well, in light of my realization above (that I can call `update()` on the HMAC multiple times). So it would look something like this?

```js
fastify.route("path", {
	preParsing(req, reply, next) => {
		const hmac = crypto.createHmac("sha1", secret);
		req.raw.on("data", d => hmac.update(d));
		req.raw.on("end", () => {
			req.digest = hmac.digest("hex");
		});
		next(); // call next immediately so the content-type parser can consume the stream
	},
	handler(req, reply) {
		if (req.raw.digest !== req.headers["x-signature"]) // invalid
	}
});
```

**Comment by mcollina** [comment]
Overall yes. You’d probably need to call .pause() as well - but it might not be needed in the basic case.

Would you like to send the relevant PRAs?

**Comment by zbjornson** [comment]
Sure, I'll give it a try (probably a week or two away from getting to it).
